tools/generate_candidate_externalids.php

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env php
+<?php
+/**
+ * Script generating an externalID for each candidate
+ * where external ID is NULL.
+ * Config useExternalID property must be set.
+ *
+ * Take an optional argument to reset all the externalID to NULL
+ * before running the algorithm
+ *
+ * Ways to use this script:
+ * -- display a help screen
+ * generate_candidate_externalids.php help
+ *
+ * -- to fix missing external IDS on candidates
+ * generate_candidate_externalids.php
+ *
+ * -- to reset all candidates external IDs and regenerate them
+ * generate_candidate_externalids.php reset
+ *
+ * PHP Version 7
+ *
+ * @category Main
+ * @package  Loris
+ * @author   Laetitia Fesselier <laetitia.fesselier@mcin.ca>
+ * @license  Loris license
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+require_once __DIR__ . "/../tools/generic_includes.php";
+
+$reset = false;
+if ((isset($argv[1]) && $argv[1] === "reset")) {
+    $reset = true;
+}
+
+/**
+ * HELP SCREEN
+ * display and stop processing if action=help
+ */
+if ((isset($argv[1]) && $argv[1] == 'help')) {
+    fwrite(STDERR, "Usage: \n\n");
+    fwrite(STDERR, "Display a help screen:\n");
+    fwrite(STDERR, "generate_candidate_externalids.php help\n\n");
+
+    fwrite(STDERR, "To fix missing external IDS on candidates:\n");
+    fwrite(STDERR, "generate_candidate_externalids.php\n\n");
+
+    fwrite(STDERR, "To reset all candidates external IDs and regenerate them:\n");
+    fwrite(STDERR, "generate_candidate_externalids.php reset\n\n");
+
+    return;
+}
+
+echo "\n---------------------------------------------------------------\n\n".
+    "  This script will generate an externalID for each candidate \n".
+    "  with misssing external ID.\n".
+    "\n---------------------------------------------------------------\n\n";
+
+if ($config->getSetting('useExternalID') !== 'true') {
+    echo "ERROR: useExternalID is deactivated in your config. \n"
+         . "Please activate it before running this script.\n\n";
+    exit(1);
+}
+
+// Get all Candidate with a NULL externalID
+$cands = $DB->pselect(
+    "SELECT CandID, ExternalID, RegistrationCenterID as site,
+    RegistrationProjectID as project
+    FROM candidate",
+    []
+);
+
+$i = 0;
+// Update all candidates with an external ID
+foreach ($cands as $cand) {
+    if (!$reset && $cand['ExternalID'] != null) {
```

Comment:
you are using a loose operator here `!=` instead of `!==` which means that this statement will return true for `$cand['ExternalID'] = false`, `$cand['ExternalID'] = 0`, `$cand['ExternalID'] = ""` or `$cand['ExternalID'] = NULL`. if this is intentional, I usually recommend using `empty()` to be more clear or making sure to add a specific comment justifying the `!=`.  If this is accidental, please use `is_null` instead


https://www.php.net/manual/en/types.comparisons.php

Reply:
Thanks, it's a typo.